Subject: PS3 Bluetooth quirk for HCI_FLT_CLEAR_ALL

This patch is still WIP.

It seems the PS3 built-in bluetooth host controler doesn't support the
HCI_FLT_CLEAR_ALL command properly.  Add a check to conditionally use
the HCI_FLT_CONN_SETUP command to clear the event filter on PS3.

With this patch bluetooth mouse or keyboard are automatically re-connected
after reboot.

Signed-off-by: Geoff Levand <geoffrey.levand@am.sony.com>

---
 include/net/bluetooth/hci.h |    6 ++++++
 net/bluetooth/hci_core.c    |   19 +++++++++++++++++++
 2 files changed, 25 insertions(+)

--- ps3-linux-dev.orig/include/net/bluetooth/hci.h
+++ ps3-linux-dev/include/net/bluetooth/hci.h
@@ -251,6 +251,12 @@ struct hci_cp_set_event_flt {
 	__u8     condition[0];
 } __attribute__ ((packed));
 
+struct hci_cp_set_event_flt_conn {
+	__u8 flt_type;
+	__u8 cond_type;
+	__u8 condition;
+} __attribute__ ((packed));
+
 /* Filter types */
 #define HCI_FLT_CLEAR_ALL	0x00
 #define HCI_FLT_INQ_RESULT	0x01
--- ps3-linux-dev.orig/net/bluetooth/hci_core.c
+++ ps3-linux-dev/net/bluetooth/hci_core.c
@@ -46,6 +46,9 @@
 
 #include <net/bluetooth/bluetooth.h>
 #include <net/bluetooth/hci_core.h>
+#if defined(CONFIG_PPC_PS3)
+#include <asm/firmware.h>
+#endif
 
 #ifndef CONFIG_BT_HCI_CORE_DEBUG
 #undef  BT_DBG
@@ -239,6 +242,22 @@ static void hci_init_req(struct hci_dev 
 		hci_send_cmd(hdev, OGF_HOST_CTL, OCF_SET_EVENT_FLT, sizeof(cp), &cp);
 	}
 
+#if defined(CONFIG_PPC_PS3)
+	/*
+	 * The PS3 built-in bluetooth host controler doesn't support the
+	 * HCI_FLT_CLEAR_ALL command properly.  Use the HCI_FLT_CONN_SETUP
+	 * command to clear the event filter.
+	 */
+	if (firmware_has_feature(FW_FEATURE_PS3_LV1)) {
+		struct hci_cp_set_event_flt_conn cp;
+		cp.flt_type = HCI_FLT_CONN_SETUP;
+		cp.cond_type = 0; /* all devices */
+		cp.condition = 1; /* auto accept is off */
+		hci_send_cmd(hdev, OGF_HOST_CTL, OCF_SET_EVENT_FLT, sizeof(cp),
+			&cp);
+	}
+#endif
+
 	/* Page timeout ~20 secs */
 	param = cpu_to_le16(0x8000);
 	hci_send_cmd(hdev, OGF_HOST_CTL, OCF_WRITE_PG_TIMEOUT, 2, &param);
